File: Algorithms and Data Structure/go/algorithms/sorting/quick-sort/quicksort.go
package quick

func sort(arr []int) {
	if 0 < len(arr) {
		pivot := partition(arr)
		sort(arr[:pivot])
		sort(arr[pivot+1:])
	}
}

func partition(arr []int) int {
	pivot := arr[len(arr)-1]
	i := -1

	for j := 0; j < len(arr)-1; j++ {
		if arr[j] < pivot {
			i++
			arr[i], arr[j] = arr[j], arr[i]
		}
	}
	i++
	arr[i], pivot = pivot, arr[i]
	return i
}

File: Custom/python/algorithms/sorts/quicksort.py
def quicksort(arr):
    _quicksort(arr, 0, len(arr)-1)

def _quicksort(arr, start, end):
    if start < end:
        pivot = partition(arr, start, end)

        _quicksort(arr, start, pivot-1)
        _quicksort(arr, pivot+1, end)


def partition(arr, start, end):
    pivot = arr[end]
    wall = start-1

    for i in range(start, end):
        if arr[i] < pivot:
            wall += 1
            arr[i], arr[wall] = arr[wall], arr[i]
        
    wall += 1
    arr[wall], pivot = pivot, arr[wall]

    return wall




test1 = [5,1,7,3,8,2,5,3,7,9]
print(test1)

quicksort(test1)
print(test1)
